WireDomain/Sources/WireDomain/Repositories/ConversationLabelsRepository.swift

```diff
@@ -0,0 +1,152 @@
+//
+// Wire
+// Copyright (C) 2024 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import CoreData
+import WireAPI
+import WireDataModel
+
+/// Facilitate access to conversation labels related domain objects.
+
+protocol ConversationLabelsRepositoryProtocol {
+
+    /// Pull conversation labels from the server and store locally
+
+    func pullConversationLabels() async throws
+}
+
+final class ConversationLabelsRepository: ConversationLabelsRepositoryProtocol {
+
+    private let userPropertiesAPI: any UserPropertiesAPI
+    private let context: NSManagedObjectContext
+
+    init(
+        userPropertiesAPI: any UserPropertiesAPI,
+        context: NSManagedObjectContext
+    ) {
+        self.userPropertiesAPI = userPropertiesAPI
+        self.context = context
+    }
+
+    /// Retrieve from backend and store conversation labels locally
+
+    func pullConversationLabels() async throws {
```

Comment:
yes that's actually a good point, repositories..writing and accessing data in specific contexts are a good candidate to be actors, no ? We could give it a try. What do you think @johnxnguyen ? (right now repositories are classes)

Reply:
I agree that repositories could be good candidates for `actor`s. I don't think an `actor` would help in this specific case because they are [reentrant](https://www.donnywals.com/actor-reentrancy-in-swift-explained/)

Reply:
great article. the provided example to handle actor reentrancy is straightforward, by keeping the state of the task but then I don't see the point of using an actor if we still need that kind of mechanism for async/await method. We can just use the same mechanism with a class and avoid race condition. So in our specific case, as you mentioned, an actor seems useless, but the 'lock' mechanism from the article is good.

Reply:
like maybe we could use this mechanism and introduce some kind of queues : an array of task with a FIFO system

Reply:
actually I don't think there's a need to do that in the case you mentioned, if `pullConversationLabels` is called twice, I guess it will just wait for the first call to finish before doing the second one since the method is async. So there would not be any race condition in that case, what do you think?

Reply:
```
// Starts first async work
let labels = await repository.pullConversationLabels() 
// Finishes first async work

// Starts second async work
let labels2 = await repository.pullConversationLabels() 
// Finishes second async work

``` 

There would not be any race condition in that case, however if we were using `async let`, these calls would run concurrently and there might be a race condition.

Reply:
yes, re-entrancy not a problem here since this is not an actor. However, as mentioned above, if were running two concurrent `pullConversationLabels` using `async let` there might be one but I don't see the cases where it would be needed to do so in the code.

Reply:
> I'm also wondering how exactly an actor will help here? What's the data race we would prevent?

The possible data race is this:

1. Call `pullConversationLabels`.
2. Labels change on backend.
3. Call `pullConversationLabels`
4. Call in `3` finishes.
5. Call in `1` finishes.

In this situation the labels now reflect the old version and not the latest version. 

This might be a very rare situation in practice and whether this is a problem or not dependents on how use `pullConversationLabels`.

For example we call `pullConversationLabels` often then it doesn't matter. We will get the correct values later anyway. But if we call `pullConversationLabels` rarely in response to a notification from backend for example then it is something we should consider.

I think the reason why I raised this concern is that the repositories API doesn't return anything, rather it updates local data as a side affect, and it feels funny that we can get in a situation where the more recent call to `pullConversationLabels` results could be override.

Personally I would consider using `Task` cancellation and have `pullConversationLabels` cancel an existing call if there is one if this is not too complex.

Reply:
I don't think it would be too complicated to implement but we should consider applying this mechanism to other calls if we want to be consistent in our codebase. We are pulling and storing in many places.

Reply:
I tried playing around with TaskGroup cancellation, it's not as easy as I thought tbh, I used `addTaskUnlessCanceled` and `cancelAll()` but things became messy, so then I figured wrapping it into a single Task and leveraging a `Task.checkCancellation()` (like the code below) to make it more straightforward, seems OK when doing some unit testing although tests are sometimes inconsistent (because we're using concurrent calls).. So I'm not 100% confident pushing this code to the branch, maybe we could leave it as is for now and keep in mind that a task cancellation mechanism may be needed if the future. Any suggestions are welcome.

```
    func pullConversationLabels() async throws {
        let conversationLabels = try await userPropertiesAPI.getLabels()
        
        ongoingTask?.cancel()
        
        ongoingTask = Task {
            let conversationLabels = try await userPropertiesAPI.getLabels()
            
            try await withThrowingTaskGroup(of: Void.self) { [self] taskGroup in
                try Task.checkCancellation()
                
                for conversationLabel in conversationLabels {
                    taskGroup.addTask { [self] in
                        try await storeLabelLocally(conversationLabel)
                    }
                }
                
                try await taskGroup.waitForAll()
            }
            
            try await deleteOldLabelsLocally(excludedLabels: conversationLabels)
        }
        
        try await ongoingTask?.value
    }
```

Reply:
👍 Thanks for investigating. OK for now I suggest we don't let this issue block the PR. Instead we can discuss with @johnxnguyen to see if this will be an actual problem in practice.